tuf/ngclient/updater.py

```diff
@@ -0,0 +1,575 @@
+                if exception.status_code not in {403, 404}:
+                    raise
+                # 404/403 means current root is newest available
+                break
+
+        # Verify final root
+        self._bundle.root_update_finished()
+
+    def _load_timestamp(self) -> None:
+        """Load local and remote timestamp metadata"""
+        try:
+            data = self._load_local_metadata("timestamp")
+            self._bundle.update_timestamp(data)
+        except (OSError, exceptions.RepositoryError) as e:
+            # Local timestamp does not exist or is invalid
+            logger.debug("Failed to load local timestamp %s", e)
+
+        # Load from remote (whether local load succeeded or not)
+        data = self._download_metadata(
+            "timestamp", DEFAULT_TIMESTAMP_MAX_LENGTH
+        )
+        self._bundle.update_timestamp(data)
+        self._persist_metadata("timestamp", data)
+
+    def _load_snapshot(self) -> None:
+        """Load local (and if needed remote) snapshot metadata"""
+        try:
+            data = self._load_local_metadata("snapshot")
+            self._bundle.update_snapshot(data)
+            logger.debug("Local snapshot is valid: not downloading new one")
+        except (OSError, exceptions.RepositoryError) as e:
+            # Local snapshot does not exist or is invalid: update from remote
+            logger.debug("Failed to load local snapshot %s", e)
+
+            metainfo = self._bundle.timestamp.signed.meta["snapshot.json"]
+            length = metainfo.length or DEFAULT_SNAPSHOT_MAX_LENGTH
+            version = None
+            if self._bundle.root.signed.consistent_snapshot:
+                version = metainfo.version
+
+            data = self._download_metadata("snapshot", length, version)
+            self._bundle.update_snapshot(data)
+            self._persist_metadata("snapshot", data)
+
+    def _load_targets(self, role: str, parent_role: str) -> None:
+        """Load local (and if needed remote) metadata for 'role'."""
+        try:
+            data = self._load_local_metadata(role)
+            self._bundle.update_delegated_targets(data, role, parent_role)
+            logger.debug("Local %s is valid: not downloading new one", role)
+        except (OSError, exceptions.RepositoryError) as e:
+            # Local 'role' does not exist or is invalid: update from remote
+            logger.debug("Failed to load local %s: %s", role, e)
+
+            metainfo = self._bundle.snapshot.signed.meta[f"{role}.json"]
+            length = metainfo.length or DEFAULT_TARGETS_MAX_LENGTH
+            version = None
+            if self._bundle.root.signed.consistent_snapshot:
+                version = metainfo.version
+
+            data = self._download_metadata(role, length, version)
+            self._bundle.update_delegated_targets(data, role, parent_role)
+            self._persist_metadata(role, data)
+
+    def _preorder_depth_first_walk(self, target_filepath) -> Dict:
```

Comment:
Stack-based preorder DFS. Nice, but too long. Could we refactor for simplicity and to check for correctness?

Reply:
#1398 : This is still an almost direct copy-paste from current client. We've been avoiding touching it so far since it's so well isolated from the code that we did touch and also because having type annotations will make following the code so much easier.